cpp/src/arrow/util/thread_pool.h

```diff
@@ -64,14 +64,39 @@ namespace detail {
 // Make sure that both functions returning T and Result<T> can be called
 // through Executor::Submit(), and that a Future<T> is returned for both.
 template <typename T>
-struct ExecutorResultTraits {
-  using ValueType = T;
+struct ExecutorSubmitTraits {
+  using FutureType = Future<T>;
 };
-
 template <typename T>
-struct ExecutorResultTraits<Result<T>> {
-  using ValueType = T;
+struct ExecutorSubmitTraits<Result<T>> {
+  using FutureType = Future<T>;
+};
+
+// Make sure that both functions returning Status and void can be called
+// through Executor::Submit(), and that a Future<> is returned for both.
+template <>
+struct ExecutorSubmitTraits<Status> {
+  using FutureType = Future<>;
 };
+template <>
+struct ExecutorSubmitTraits<void> {
+  using FutureType = Future<>;
+};
+
+template <typename T, typename F>
+typename std::enable_if<
+    !std::is_same<typename std::result_of<F && ()>::type, void>::value>::type
+ExecuteAndMarkFinished(Future<T>* fut, F&& f) {
```

Comment:
Is it actually better to define this here rather than inside `Future`?

Reply:
Since it's only used in Executor::Submit it seemed preferable to simplify the public interface of `Future` by moving it here. It can certainly be moved back if there's another application for it

Reply:
It seems like this can be a generic primitive on futures (which generally entail the execution of a function-like before being mark finished).

Reply:
> which generally entail the execution of a function-like before being mark finished

The only use of this in our code was in Executor::Submit, so I'm not sure what you're referring to

Reply:
I'm not saying it's required currently to expose it, just that it sounds potentially useful. But, well, we can move it back later...